platform/osgi/platform-osgi-configadmin/src/main/java/org/codice/felix/cm/file/DelegatingPersistenceManager.java

```diff
@@ -0,0 +1,112 @@
+import org.codice.felix.cm.internal.ConfigurationContextFactory;
+import org.codice.felix.cm.internal.ConfigurationPersistencePlugin;
+import org.osgi.framework.BundleContext;
+import org.osgi.framework.FrameworkUtil;
+import org.osgi.framework.ServiceRegistration;
+import org.osgi.util.tracker.ServiceTracker;
+
+/**
+ * The {@link DelegatingPersistenceManager} is responsible for iterating over relevant configuration
+ * plugin points prior to moving the configs to the inner {@link PersistenceManager}.
+ */
+public class DelegatingPersistenceManager extends WrappedPersistenceManager {
+
+  private final ServiceTracker configPersistenceTracker;
+
+  private final ServiceRegistration<ConfigurationContextFactory> contextFactoryRegistration;
+
+  public DelegatingPersistenceManager(PersistenceManager persistenceManager) {
+    this(
+        persistenceManager,
+        new ServiceTracker(getBundleContext(), ConfigurationPersistencePlugin.class, null),
+        getBundleContext()
+            .registerService(
+                ConfigurationContextFactory.class, new ConfigurationContextFactoryImpl(), null));
+  }
+
+  DelegatingPersistenceManager(
+      PersistenceManager persistenceManager,
+      ServiceTracker tracker,
+      ServiceRegistration<ConfigurationContextFactory> registration) {
+    super(persistenceManager);
+    tracker.open();
+
+    this.configPersistenceTracker = tracker;
+    this.contextFactoryRegistration = registration;
+  }
+
+  @Override
+  public void close() throws Exception {
+    configPersistenceTracker.close();
+    if (contextFactoryRegistration != null) {
+      contextFactoryRegistration.unregister();
+    }
+    super.close();
+  }
+
+  @Override
+  public void store(String pid, Dictionary properties) throws IOException {
+    ConfigurationContextImpl context = createContext(pid, properties);
+    if (context.shouldBeVisibleForProcessing()) {
+      List<ConfigurationPersistencePlugin> plugins =
+          asPlugins(configPersistenceTracker.getServices());
+      for (ConfigurationPersistencePlugin plugin : plugins) {
+        plugin.handleStore(context);
+      }
+    }
+    super.store(pid, properties);
+  }
+
+  @Override
+  public void delete(String pid) throws IOException {
+    List<ConfigurationPersistencePlugin> plugins =
+        asPlugins(configPersistenceTracker.getServices());
+    for (ConfigurationPersistencePlugin plugin : plugins) {
+      plugin.handleDelete(pid);
+    }
+    super.delete(pid);
+  }
+
+  /* Factory method visible for testing purposes only */
+  ConfigurationContextImpl createContext(String pid, Dictionary props) {
+    return new ConfigurationContextImpl(pid, props);
+  }
+
+  private static List<ConfigurationPersistencePlugin> asPlugins(Object[] serviceObjects) {
+    if (serviceObjects == null) {
+      return Collections.emptyList();
+    }
+    List<ConfigurationPersistencePlugin> plugins = new ArrayList<>();
+    for (Object object : serviceObjects) {
```

Comment:
✏️ Could be transformed into a nice Stream.of(serviceObjects) .... collect()

Reply:
❗️  In our custom config admin, we are using an older version of the maven bundle plugin to match how Felix does things. There is a bug in this version that will cause the build to fail on lambdas and other elegant Java 8 constructs:
https://issues.apache.org/jira/browse/FELIX-4005
Also, even if we jump ahead, we have to be careful. Because the issue reappeared and had to be fixed again:
https://issues.apache.org/jira/browse/FELIX-4556

Reply:
:-( I remember now. Maybe adding a comment at the top of the files that are affected by this would be good for the next developer that comes here to fix something.